cmd/agent/app/snmpwalk.go

```diff
@@ -0,0 +1,316 @@
+	communityString string
+
+	// v3
+	user          string
+	authProtocol  gosnmp.SnmpV3AuthProtocol
+	authProt      string
+	authKey       string
+	privProtocol  gosnmp.SnmpV3PrivProtocol
+	privProt      string
+	privKey       string
+	snmpContext   string
+	msgFlags      gosnmp.SnmpV3MsgFlags
+	securityLevel string
+
+	// communication
+	retries int
+	timeout int
+)
+
+var (
+	snmpCmd = &cobra.Command{
+		Use:   "snmp",
+		Short: "Snmp tools",
+		Long:  ``,
+	}
+)
+
+func init() {
+
+	snmpWalkCmd.Flags().StringVarP(&snmpVersion, "snmp-version", "v", defaultVersion, "Specify SNMP version to use")
+
+	// snmp v1 or v2c specific
+	snmpWalkCmd.Flags().StringVarP(&communityString, "community-string", "C", "", "Set the community string")
+
+	// snmp v3 specific
+	snmpWalkCmd.Flags().StringVarP(&authProt, "auth-protocol", "a", defaultAuthProtocol, "Set authentication protocol (MD5|SHA|SHA-224|SHA-256|SHA-384|SHA-512)")
+	snmpWalkCmd.Flags().StringVarP(&authKey, "auth-key", "A", defaultAuthKey, "Set authentication protocol pass phrase")
+	snmpWalkCmd.Flags().StringVarP(&securityLevel, "security-level", "l", defaultSecurityLevel, "set security level (noAuthNoPriv|authNoPriv|authPriv)")
+	snmpWalkCmd.Flags().StringVarP(&snmpContext, "context", "N", defaultContext, "Set context name")
+	snmpWalkCmd.Flags().StringVarP(&user, "user-name", "u", defaultUserName, "Set security name")
+	snmpWalkCmd.Flags().StringVarP(&privProt, "priv-protocol", "x", defaultPrivProtocol, "Set privacy protocol (DES|AES|AES192|AES192C|AES256|AES256C)")
+	snmpWalkCmd.Flags().StringVarP(&privKey, "priv-key", "X", defaultPrivKey, "Set privacy protocol pass phrase")
+
+	// general communication options
+	snmpWalkCmd.Flags().IntVarP(&retries, "retries", "r", defaultRetries, "Set the number of retries")
+	snmpWalkCmd.Flags().IntVarP(&timeout, "timeout", "t", defaultTimeout, "Set the request timeout (in seconds)")
+
+	snmpWalkCmd.SetArgs([]string{})
+
+	// attach snmpWalk to snmp command
+	snmpCmd.AddCommand(snmpWalkCmd)
+
+	// attach the command to the root
+	AgentCmd.AddCommand(snmpCmd)
+}
+
+var snmpWalkCmd = &cobra.Command{
+	Use:   "walk <IP Address>[:Port] [OID] [OPTIONS]",
+	Short: "Perform a snmpwalk",
+	Long:  ``,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		// Get args
+		if len(args) == 0 {
+			fmt.Print("Missing argument: IP address\n")
+			cmd.Help() //nolint:errcheck
+			os.Exit(1)
+			return nil
+		} else if len(args) == 1 {
+			address = args[0]
+			oid = defaultOID
+		} else if len(args) == 2 {
+			address = args[0]
+			oid = args[1]
+		} else {
+			fmt.Printf("The number of arguments must be between 1 and 2. %d arguments were given.\n", len(args))
+			cmd.Help() //nolint:errcheck
+			os.Exit(1)
+			return nil
+		}
+		if strings.Contains(address, ":") {
+			deviceIP = address[:strings.Index(address, ":")]
+			value, _ = strconv.ParseUint(address[strings.Index(address, ":")+1:], 0, 16)
+			port = uint16(value)
+			if port == 0 {
+				port = defaultPort
+			}
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏If user sets `127.0.0.1:0` should we return an error instead ? I'm not sure if in this case, we should replace `0` by defaultPort since the input seems not valid.

Reply:
@AlexandreYang is port `0` a possibly valid port ?

Reply:
Looking a bit about it. I think `0` is not a valid SNMP port.

> Unlike most [port numbers](https://www.lifewire.com/port-numbers-on-computer-networks-817939), port 0 is a reserved port in [TCP/IP](https://www.lifewire.com/transmission-control-protocol-and-internet-protocol-816255) networking, meaning that it should not be used in TCP or [UDP](https://www.lifewire.com/list-of-tcp-ports-and-udp-ports-818133) messages.

https://www.lifewire.com/port-0-in-tcp-and-udp-818145

Reply:
> This choice was made to be coherent with what we already have for the [corecheck](https://github.com/DataDog/datadog-agent/blob/18df6bd481309a251d5a39a5b48274460b9a1c10/pkg/snmp/snmp.go#L117-L119)

In the case of the corecheck, the check for `0` is mainly to check if the port is not set.

What I think is invalid is when user setting explicitely port to zero like `127.0.0.1:0`. I think for this case, we should issue an error instead of replacing `0` by default port.